r/src/altrep.cpp

```diff
@@ -44,25 +49,58 @@ extern "C" {
 #endif
 
 #include <arrow/array.h>
+#include <arrow/util/bitmap_reader.h>
+
+#include "./r_task_group.h"
 
 namespace arrow {
 namespace r {
 
+template <typename T>
+void UseSentinel(const std::shared_ptr<Array>& array) {
```

Comment:
I would have expected that, instead of altering the Arrow data, we would have `_ELT` and `_GET_REGION` methods that would insert the sentinels in the R vector when the data is accessed. What are the tradeoffs of that way vs. this way?

Reply:
Oh interesting. I guess the advantage of this approach is that, for the price of this initial setting of the sentinels, we can then have a `DATAPTR()` so access all the data of the array randomly. 

OTOH, we could leave the data untouched and have an `_Elt` that would check the null bitmap, and an `GetRegion()` that would also locally use the null bitmap. This `GetRegion()` would indeed copy the region data though, so potentially be more expensive, even when the region has no nulls.

Reply:
Chances are we might need to implement something like this for arrays that are backed by an immutable buffer, so maybe I can do that, and then we can compare/measure and evaluate how risky this is. 

Then we have 3 cases: 
 - Array with no nulls: best altrep with DATAPTR(), no need for setting sentinels (what we had prior to this pull request). 
 - Array with some nulls, immutable: using GetRegion, I guess with a materialized R vector in the data2 altrep slot, so that DATAPTR() materializes and returns this. 
 
 - Array with some nulls, "mutable": hacky proposal of this pr. 

Perhaps we only need the first 2.

Reply:
Also, I believe the second dot will have some code shared with altrep for ChunkedArray.